### ISSUE TITLE
Add support for scanning postgresql `interval` type into `time.Duration`

### ISSUE THREAD

**Issue by stevvooe** [lib/pq#78]
Just thought I'd file this as a nice to have.

**Comment by cdipaolo** [comment]
Not sure if this is still up for debate but it would be nice to have in my use case instead of having to convert the string to a duration by hand in go. 

:+1:

**Comment by johto** [comment]
The point still stands: you can't convert an interval to a Duration without losing information.  What was said above only touches the surface: postgres' intervals don't store just seconds, but also the number of days and months.  So you could have an interval of "1 month 1 day 34 seconds", which doesn't map into a time.Duration at all.

**Comment by cdipaolo** [comment]
Yeah after further research yesterday I found that out. You could
potentially map to a custom struct but that wouldn't be very idiomatic or
clean

**Comment by stevvooe** [comment]
Perhaps, we could re-specify this issue and have `lib/pq` provide common time types that interoperate with the postgresql timestamp and interval types. The main issue right now is that its hard to map any postgresql timestamp or interval types. This is a massive gap in functionality and hard to get right for each individual project.

**Comment by johto** [comment]
> Perhaps, we could re-specify this issue and have lib/pq provide common time types that interoperate with the postgresql timestamp and interval types. The main issue right now is that its hard to map any postgresql timestamp or interval types. This is a massive gap in functionality and hard to get right for each individual project.

I think what we could do is provide a struct which stores the values for all the individual components in an interval.  You could then access said components through the struct or you could ask for the struct to convert itself into a time.Duration using the same rules `extract(epoch from interval)` in postgres does.  It shouldn't be more than a SMOP for someone to implement.

**Comment by AlekSi** [comment]
I think this issue should be renamed – as pointed above it's not possible to map `interval` to `time.Duration`.

**Comment by stevvooe** [comment]
@johto Your suggestion sounds reasonable, as it will make specialized field mapping easier to work with.

@AlekSi The edge case here can be detected as an error. It doesn't make sense to omit support for `interval` with `time.Duration` when the vast majority of use cases don't need +/- 178000000. If the `interval` overflows `time.Duration`, we could simply throw an error. Another type can be provided in `lib/pq` for use in cases where ranges larger than +/- 292 years are actually required.

**Comment by kostya-sh** [comment]
@stevvooe `interval` and `time.Duration` are incompatible types. `time.Duration` represents duration between two moments in time, while `interval` represents interval between two dates. 1 day interval is not the same as 24 hours duration: http://play.golang.org/p/OLF9L2EsX6

**Comment by stevvooe** [comment]
@kostya-sh Thanks for bringing up the difference between calendrical calculations and exact differentials. Just as there may be overflow of the `time.Duration` type, there may be intervals, such as those with components that need an anchor to evaluate, that don't map to `time.Duration`.

However, there are entire classes of `interval` values that are completely compatible. For example, let's take this:

```console
postgres=# select extract(epoch from i), extract(days from i) from (select interval '1000000 seconds' as i) as foo;
 date_part | date_part
-----------+-----------
   1000000 |         0
(1 row)
```

As we can see, the seconds component of this interval is more than compatible with `time.Duration`. There are no days or years.

But, yes, I agree that `interval` expressed in terms of days is incompatible:

```console
postgres=# select extract(epoch from i), extract(days from i), extract(years from i) from (select interval '1 year>
 date_part | date_part | date_part
-----------+-----------+-----------
  33544800 |        23 |         1
(1 row)
```

This would have to be mapped to a type that exposes the calendrical components, if required. Oddly enough, the use of `epoch` does evaluate the `interval` without binding it to a date, which could be compatible with `time.Duration`.

My point is that this isn't quite as absolute as claimed. The overlap in domains of `time.Duration` and `interval` tends to be useful in practice such that omitting support for `time.Duration` hurts the usability of this package. As it stands now, there is now way to extract this from `pq`, which is disappointing.

At least, let's take the first step of having an `interval` type.

**Comment by kostya-sh** [comment]
I think adding `pq.Interval` type is the best solution here. I believe it is reasonable for a driver package to provide Go types for database specific data types.

**Comment by ear** [comment]
Hi. I'd love to have a type to work with for Postgres Intervals. Just in case you decide not to put it in lib/pq can you give an example of how to implement it in user code if at all possible? Thank you.

**Comment by teepark** [comment]
I've started a project to tackle this with an Interval type that fully matches postgres's semantics and a Duration alias that converts according to the EXTRACT(EPOCH) behavior. Total WIP at the moment but it's starting to successfully scan intervals.

https://github.com/teepark/pqinterval

**Comment by icholy** [comment]
Since `time.Duration` is an `int64`, why not simply map it to a `bigint`?

**Comment by aseemk** [comment]
@teepark: Thank you for your library! Fantastic test coverage. ♥️

**Comment by TheWug** [comment]
I understand that the two types don't line up well but I've also found myself revisiting this thread numerous times over the years and after putting some thought into it I have a proposal.

Obstacles:
The postgres `interval` type isn't simple, it contains the following:
* integral numbers of various precisions representing years, months, weeks, days, hours, minutes
* double precision floating point number of seconds
The golang `time.Duration` type is simple, it is just a signed int64 number of nanoseconds.
Some `interval` represent abstract periods of time, and `Duration` represents only concrete periods of time.

Each has large ranges of values not representable by the other.
1. The duration's range is +/-290 years, and the interval's is +/-178M years
2. The interval's double precision seconds field is only nanosecond-accurate up to about 52 days

Proposal: Provide a `pq.Interval` class which exactly represents `interval`, and allow conversions between it and `time.Duration` relative to a particular `time.Time`. `interval` seconds isn't precise enough to represent all `Duration`, but if you calculate relative to a given reference date, it's possible to exactly calculate an `interval` from any `Duration` by making use of the other fields. Likewise, when going in the other direction, ambiguous `interval` like "1 year" resolve to an exact period of time if they have a reference date. Because the domain of `interval` is larger, there will still be intervals that `Duration` cannot represent, but these can be handled by raising a domain error.

Having a reference date achieves the following important property:
```var original time.Duration
var t time.Time
final, err := pq.ToDuration(pq.ToInterval(d, t), t) // err == nil, and original == final
```

In other words, any `time.Duration` can be reversibly converted into a postgres `interval`.  It's not an ideal solution but I think it's an acceptable compromise. Clients which need a greater domain than the one `Duration` allows can either directly examine the `pq.Interval` object or use the database itself to perform date math.

**Comment by jordan-bonecutter** [comment]
Just want to throw this out there: I can scan a bigint into an int8. In general, the sql libs are not super concerned with data sizing so not doing this for data sizing reasons seems inconsistent at best. As a rule of thumb for sql, if you need data to be exactly as it is in the db, parse into a string or []byte. I think this decision should be reconsidered, it will fix more bugs than it will create (think of all the crappy ParseDuration functions people like me are about to write because of this!)

**Comment by teepark** [comment]
I never followed up here again years ago, but [pqinterval](https://github.com/teepark/pqinterval) worked well in production for years. Going back through the discussion here, it seems like the approach I took (matching "extract(epoch)" semantics) was the only agreed-upon reasonable way to get a postgresql interval into a time.Duration. For cases where those specific semantics aren't desired, or the result would overflow a Duration, the package exposes a Scanner- and Valuer-compatible Interval type as well.

@fdr would you entertain a PR that brings pqinterval into lib/pq?
